### PR TITLE
speed up ts

### PR DESCRIPTION
**Pull request by chrisnojima**
My process to debug this:
run
`time yarn tsc --diagnostics --extendedDiagnostics --declarationDir ~/Desktop/decl`

then look at the output d.ts files and see where types are exploding. Manually fix them up.

tldr; a full TS run goes from 164s => 87s

master:
```
Files:                        2503
Lines:                      369589
Nodes:                     1781690
Identifiers:                567804
Symbols:                   1095193
Types:                      449112
Memory used:              1742305K
Assignability cache size:  1489182
Identity cache size:       1299331
Subtype cache size:        1057202
I/O Read time:               0.19s
Parse time:                  1.60s
Program time:                2.87s
Bind time:                   1.30s
Check time:                 77.07s
transformTime time:         66.50s
commentTime time:            3.79s
printTime time:             83.21s
Emit time:                  83.27s
Source Map time:             1.97s
I/O Write time:              0.71s
Total time:                164.52s
```

PR:
```
Files:                        2512
Lines:                      371110
Nodes:                     1784914
Identifiers:                568033
Symbols:                   1111845
Types:                      451149
Memory used:              1304213K
Assignability cache size:  1489577
Identity cache size:       1299346
Subtype cache size:         759405
I/O Read time:               0.19s
Parse time:                  1.62s
Program time:                2.93s
Bind time:                   1.25s
Check time:                 69.61s
transformTime time:         10.78s
commentTime time:            0.71s
printTime time:             13.96s
Emit time:                  13.99s
I/O Write time:              0.36s
Total time:                 87.77s
```

### CODE REVIEW

File: shared/actions/fs/shared.tsx [hunk 1]
```diff
@@ -13,11 +12,9 @@ const expectedOfflineErrorMatchers = [
 
 const expectedOfflineErrorCodes = [RPCTypes.StatusCode.scapinetworkerror, RPCTypes.StatusCode.scstreameof]
 
-const makeErrorHandler = (
-  action: FsGen.Actions | EngineGen.Actions,
```
Comment (chrisnojima) [inline shared/actions/fs/shared.tsx]: merging FsGen and EngineGen makes a very big type, just say all actions can go

File: shared/actions/platform-specific/index.desktop.tsx [hunk 2]
```diff
@@ -85,7 +85,7 @@ function* initializeInputMonitor(): Iterable<any> {
   }
 }
 
-export const dumpLogs = async (_?: Container.TypedState, action?: ConfigGen.DumpLogsPayload) => {
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.desktop.tsx]: don't need this

File: shared/chat/conversation/container.tsx [hunk 3]
```diff
@@ -22,7 +22,7 @@ type SwitchProps = {
   type: 'error' | 'noConvo' | 'rekey' | 'youAreReset' | 'normal' | 'rekey'
 }
 
-class Conversation extends React.PureComponent<SwitchProps> {
+export class Conversation extends React.PureComponent<SwitchProps> {
```
Comment (chrisnojima) [inline shared/chat/conversation/container.tsx]: if we don't export it, it has to copy a bunch of stuff to other files

File: shared/chat/conversation/messages/reactions-row/index.tsx [hunk 4]
```diff
@@ -45,8 +45,8 @@ class ReactionsRow extends React.Component<Props, State> {
     this.setState(s => (s.showMobileTooltip === showMobileTooltip ? null : {showMobileTooltip}))
 
   _newAttachmentRef: typeof ReactButton | null = null
-  _getNewAttachmentRef = () => this._newAttachmentRef
-  _setNewAttachmentRef = (r: typeof ReactButton) => {
+  private _getNewAttachmentRef = () => this._newAttachmentRef
```
Comment (chrisnojima) [inline shared/chat/conversation/messages/reactions-row/index.tsx]: else it exports and needs a bunch of stuff

File: shared/chat/payments/status/index.tsx [hunk 5]
```diff
@@ -93,7 +93,7 @@ class PaymentStatus extends React.Component<Props, State> {
           type={getIcon(this.props.status)}
           fontSize={12}
           boxStyle={styles.iconBoxStyle}
-          style={styles[statusIcon(this.props.status)]}
+          color={statusColor(this.props.status)}
```
Comment (chrisnojima) [inline shared/chat/payments/status/index.tsx]: this was a real bug, color has to be passed in separately

File: shared/constants/fs.tsx [hunk 6]
```diff
@@ -89,7 +89,11 @@ export const tlfSyncDisabled: Types.TlfSyncDisabled = {
   mode: Types.TlfSyncMode.Disabled,
 }
 
-export const makeTlfSyncPartial = ({enabledPaths}: Partial<Types.TlfSyncPartial>): Types.TlfSyncPartial => ({
+export const makeTlfSyncPartial = ({
```
Comment (chrisnojima) [inline shared/constants/fs.tsx]: i did this as an experiment, if you're just pulling out certain fields, its faster to just pull them out

File: shared/constants/types/autoreset.tsx [hunk 7]
```diff
@@ -1,8 +1,8 @@
-export type State = Readonly<{
```
Comment (chrisnojima) [inline shared/constants/types/autoreset.tsx]: this is a bummer. Readonly<> is way slower than making everything readonly

File: shared/desktop/yarn-helper/font.tsx [hunk 8]
```diff
@@ -380,7 +380,12 @@ function updateIconConstants() {
     requireDark?: string
   }
 
-  const iconMeta_ = {
+  export type IconType = ${Object.keys(icons)
```
Comment (chrisnojima) [inline shared/desktop/yarn-helper/font.tsx]: makes the generated type way faster

File: shared/reducers/index.tsx [hunk 9]
```diff
@@ -51,8 +51,35 @@ const reducers = {
   wallets,
 }
 
-const rootReducer = combineReducers(reducers)
-export type TypedState = ReturnType<typeof rootReducer>
+// this is to help TS make this type more static so it emits cleanly
+const rootReducer: any = combineReducers(reducers as any)
+
+export type TypedState = {
```
Comment (chrisnojima) [inline shared/reducers/index.tsx]: also a bummer but ensures this type doesnt' get copied everywhere

File: shared/reducers/tracker2.tsx [hunk 10]
```diff
@@ -97,8 +97,7 @@ export default Container.makeReducer<Actions, Types.State>(initialState, {
     d.followingCount = d.following.size
   },
   [Tracker2Gen.proofSuggestionsUpdated]: (draftState, action) => {
-    type ReadonlyProofSuggestions = Readonly<Types.State['proofSuggestions']>
-    ;(draftState.proofSuggestions as ReadonlyProofSuggestions) = action.payload.suggestions
+    draftState.proofSuggestions = Container.castDraft(action.payload.suggestions)
```
Comment (chrisnojima) [inline shared/reducers/tracker2.tsx]: immer has new helpers for this now

File: shared/styles/colors.tsx [hunk 11]
```diff
@@ -397,7 +397,7 @@ export const darkColors: {[P in keyof typeof colors]: string | undefined} = {
   },
 }
 
-const partyFallbackColors = {
+const partyFallbackColors: {[P in keyof typeof colors]?: string | undefined} = {
```
Comment (chrisnojima) [inline shared/styles/colors.tsx]: else the whole thing gets inlined

File: shared/styles/css.d.ts [hunk 12]
```diff
@@ -1,8 +1,6 @@
 import {CSSProperties} from 'react'
 import {ViewStyle, TextStyle, ImageStyle} from 'react-native'
 
-type StyleProp<T> = T | Array<StyleProp<T>> | undefined | null | false
```
Comment (chrisnojima) [inline shared/styles/css.d.ts]: similar problem to Readonly

File: shared/wallets/sep7-confirm/index.tsx [hunk 13]
```diff
@@ -134,7 +134,7 @@ const Header = (props: HeaderProps) => (
         </Kb.Text>
         {props.signed && (
           <Kb.Box2 direction="horizontal" style={styles.verifiedIconBox}>
-            <Kb.Icon sizeType="Small" style={styles.verifiedIcon} type="iconfont-success" />
+            <Kb.Icon sizeType="Small" type="iconfont-success" color={Styles.globalColors.green} />
```
Comment (chrisnojima) [inline shared/wallets/sep7-confirm/index.tsx]: also a bug!

File: shared/constants/types/autoreset.tsx (hunk 7 above)
Comment (jakob223) [reply shared/constants/types/autoreset.tsx]: 😢

File: shared/desktop/yarn-helper/font.tsx [hunk 14]
```diff
@@ -403,11 +408,8 @@ function updateIconConstants() {
         }`
       })
       .join(',\n')
-  }/* eslint-enable */
```
Comment (jakob223) [inline shared/desktop/yarn-helper/font.tsx]: did you intend to delete this?

File: shared/styles/css.d.ts (hunk 12 above)
Comment (jakob223) [reply shared/styles/css.d.ts]: this is sad

File: shared/tsconfig.base.json [hunk 15]
```diff
@@ -34,9 +36,9 @@
     "noUnusedParameters": true,
     "outDir": "./.tsOuts/.tsOut/emit",
     "resolveJsonModule": true,
-    "sourceMap": true,
+    "sourceMap": false,
```
Comment (jakob223) [inline shared/tsconfig.base.json]: This change looks scary - what does it do? will it make debugging harder?

File: shared/profile/non-user-profile/container.tsx [hunk 16]
```diff
@@ -7,7 +7,7 @@ import {privateFolderWithUsers} from '../../constants/config'
 import NonUserProfile from '.'
 import {ServiceId} from 'util/platforms'
 
-type OwnProps = Container.RouteProps<{
+export type OwnProps = Container.RouteProps<{
```
Comment (jakob223) [inline shared/profile/non-user-profile/container.tsx]: Can you add a comment saying export is for TS optimization?

File: shared/profile/pgp/finished.desktop.tsx [hunk 17]
```diff
@@ -18,7 +18,7 @@ type State = {
   shouldStoreKeyOnServer: boolean
 }
 
-class Finished extends React.Component<Props, State> {
+export class Finished extends React.Component<Props, State> {
```
Comment (jakob223) [inline shared/profile/pgp/finished.desktop.tsx]: Can you add a comment saying export is for TS optimization?

File: shared/profile/user/container.tsx [hunk 18]
```diff
@@ -7,7 +7,7 @@ import * as Types from '../../constants/types/tracker2'
 import Profile2, {BackgroundColorType} from '.'
 import flags from '../../util/feature-flags'
 
-type OwnProps = Container.RouteProps<{username: string}>
+export type OwnProps = Container.RouteProps<{username: string}>
```
Comment (jakob223) [inline shared/profile/user/container.tsx]: Can you add a comment saying export is for TS optimization?

File: shared/signup/phone-number/container.tsx [hunk 19]
```diff
@@ -14,7 +14,7 @@ type WatcherProps = Props & {
   pendingVerification: string
 }
 // Watches for `pendingVerification` to change and routes to the verification screen
-class WatchForGoToVerify extends React.Component<WatcherProps> {
+export class WatchForGoToVerify extends React.Component<WatcherProps> {
```
Comment (jakob223) [inline shared/signup/phone-number/container.tsx]: Can you add a comment saying export is for TS optimization?

File: shared/signup/phone-number/verify-container.tsx [hunk 20]
```diff
@@ -14,7 +14,7 @@ type WatcherProps = Props & {
   verificationStatus?: 'success' | 'error'
 }
 // Watches for verification to succeed and exits
-class WatchForSuccess extends React.Component<WatcherProps> {
+export class WatchForSuccess extends React.Component<WatcherProps> {
```
Comment (jakob223) [inline shared/signup/phone-number/verify-container.tsx]: Can you add a comment saying export is for TS optimization?

File: shared/desktop/yarn-helper/font.tsx (hunk 14 above)
Comment (chrisnojima) [reply shared/desktop/yarn-helper/font.tsx]: Yes it's unused

File: shared/tsconfig.base.json (hunk 15 above)
Comment (chrisnojima) [reply shared/tsconfig.base.json]: We don't use tsc for code gen

File: shared/chat/conversation/container.tsx (hunk 3 above)
Comment (chrisnojima) [reply shared/chat/conversation/container.tsx]: These individual things aren't so bad so if it comes back it's not the end of the world

**Pull request merged**